bullet-featherstone/src/Base.cc

```diff
@@ -26,14 +26,213 @@ namespace gz {
+        {
+          const btCollisionShape *shape = _compoundShape->getChildShape(i);
+          if (this->HasConvexHullChildShapes(shape))
+            return nullptr;
+        }
+      }
+
+      const btCollisionShape *shape =
+          _compoundShape->getChildShape(_childIndex);
+      return shape;
+    }
+    else
+    {
+      return _compoundShape->getChildShape(0);
+    }
+  }
+  return nullptr;
+}
+
+/////////////////////////////////////////////////
+void GzCollisionDispatcher::dispatchAllCollisionPairs(
+    btOverlappingPairCache* pairCache,
+    const btDispatcherInfo& dispatchInfo,
+    btDispatcher* dispatcher)
+{
+  btCollisionDispatcher::dispatchAllCollisionPairs(
+      pairCache, dispatchInfo, dispatcher);
+
+  // Loop through all the contact manifolds.
+  // Find convex decomposed mesh collision shapes.
+  // Create a shared contact manifold for all decomposed shapes.
+  // This is so that we can limit the number of contact points to 4
+  // for the collision shape. Otherwise it will generate up to
+  // (decomposed col count * 4) contact points.
+  int numManifolds = this->getNumManifolds();
+  for (int i = 0; i < numManifolds; ++i)
+  {
+    btPersistentManifold* contactManifold =
+        this->getManifoldByIndexInternal(i);
+
+    const btMultiBodyLinkCollider* ob0 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody0());
+    const btMultiBodyLinkCollider* ob1 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody1());
+
+    // if it's a custom manifold, just refresh contacts, no need to
+    // loop through and check them.
+    if (this->customManifolds.find(contactManifold) !=
+        this->customManifolds.end())
+    {
+      contactManifold->refreshContactPoints(ob0->getWorldTransform(),
+                                            ob1->getWorldTransform());
+      continue;
+    }
+
+    if (this->manifoldsToKeep.find(contactManifold) !=
+        this->manifoldsToKeep.end())
+    {
+      continue;
+    }
+
+    const btCompoundShape *compoundShape0 =
+        static_cast<const btCompoundShape *>(ob0->getCollisionShape());
+    const btCompoundShape *compoundShape1 =
+        static_cast<const btCompoundShape *>(ob1->getCollisionShape());
+
+    int numContacts = contactManifold->getNumContacts();
+    for (int j = 0; j < numContacts; ++j)
+    {
+      btManifoldPoint& pt = contactManifold->getContactPoint(j);
+      const btCollisionShape *colShape0 = this->FindCollisionShape(
+          compoundShape0, pt.m_index0);
+      const btCollisionShape *colShape1 = this->FindCollisionShape(
+          compoundShape1, pt.m_index1);
+
+      if (!colShape0 || !colShape1 ||
+         (!this->HasConvexHullChildShapes(colShape0) &&
+          !this->HasConvexHullChildShapes(colShape1)))
+      {
+        this->manifoldsToKeep.insert(contactManifold);
+        continue;
+      }
+
+      btPersistentManifold* colManifold =
```

Comment:
When we remove manifolds, we don't do anything with the pointer in `colPairManifolds`, so we it might not be a nullptr even if the manifold has been removed. I suggest checking if it's in `customManifolds` instead of just a `nullptr` check.

Reply:
I updated the code to remove the pointers in `colPairManifolds` when manifolds are removed. So the `colPairManifolds` map should always be up-to-date now. [15d4ede](https://github.com/gazebosim/gz-physics/pull/664/commits/15d4ede77239f9a1b35971a13a9c3fe4ccaa1bca)